### PR TITLE
Add a new EntryType for deletion with timestamp

### PR DESCRIPTION
**Pull request by riversand963**
Add `kEntryDeleteWithTimestamp` to `EntryType` which is a public API.

Test plan:
make check

### CODE REVIEW

**Comment by riversand963** [comment]
Will update HISTORY.md later.
IMO, I think `EntryType` should use safer enum types supported in C++, or at least assign values explicitly, but for now I'll stick to the current convention.

**Comment by riversand963** [comment]
cc @hliu18

**Comment by hliu18** [comment]
LGTM

File: include/rocksdb/types.h [hunk 1]
```diff
@@ -21,6 +21,7 @@ const SequenceNumber kMinUnCommittedSeq = 1;  // 0 is always committed
 enum EntryType {
   kEntryPut,
   kEntryDelete,
+  kEntryDeleteWithTimestamp,
```
Comment (ajkr) [inline include/rocksdb/types.h:24]: enum order changes kind of scare me. For example, someone might store or transfer some keys/values along with the EntryType. In that case they likely encode EntryType as an int. So upgrading, or transferring to a host running an older/newer version will decode that int as a different EntryType from what was intended.

We could comment on this enum that the ordering is subject to change.

Comment (riversand963) [reply include/rocksdb/types.h:24]: It scares me too. Imo, int-based enums should always have explicit value assignments. I looked at previous PRs changing this enum, and they also changed the order...

Comment (riversand963) [reply include/rocksdb/types.h:24]: okay, let me not change the ordering of enum.

**Pull request closed without merging**